### PR TITLE
19-implement-unsupervised-exploration

### LINKED ISSUE #9: Create playground unsupervised exploration
Create a first test implementation for unsupervised exploration (PEBBLE)

### LINKED ISSUE #19: Implement unsupervised exploration
Implement unsupervised exploration as described in https://arxiv.org/pdf/2106.05091

### PR DESCRIPTION
**Pull request by Wizzardexe**
**Issue:** Fixes #9 Fixes #19

**What?**
- I implemented the particle-based entropy estimator in unsupervised_exploration.py
- Added some modularity to sac_rhlf.py
- Added in a small effort to get better testing results 1 hidden layer
- Noticed a small problem in the preference_loss resulting in infinite loss

**How?**
- For the estimator I used a KNN
- To avoid to have log(0) in the loss I added a clamp

**Problems rightnow**
- With the current buffer implementation, I need to use a separate buffer, therefore the generated trajectories cannot be directly used for comparison. The issue arises because the buffer stores the GroundTruth, whereas it should store the reward from the reward model or the intrinsic reward instead.

### CODE REVIEW

File: framework/video_recorder.py [hunk 1]
```diff
@@ -54,6 +54,6 @@ def record_trajectory(self, trajectory: Trajectory, run_name: str, fps=30):
 
             # Save the rendered image to the video file
             writer.write(env.render())
-
+        print("video_done")
```
Comment (bkuen) [inline framework/video_recorder.py:57]: I would remove this to avoid the console being flooded. Maybe we should consider adding a better logger for debug messages like this one in the future.

File: framework/unsupervised_exploration.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        visited_states = np.array(self.visited_states)
+        visited_states = (visited_states - visited_states.mean(axis=0)) / (visited_states.std(axis=0) + 1e-8)
+        tree = KDTree(visited_states)
+
+        # Normalizing the data to have a mean of 0 and a standard deviation of 1 allows the KDTree
+        # to process all dimensions on an equal footing, irrespective of their original distributions.
+        states = (states - visited_states.mean(axis=0)) / (visited_states.std(axis=0) + 1e-8)
+
+        # Query the KDTree for distances
+        distances, _ = tree.query(states, k=self.k + 1)
```
Comment (bkuen) [inline framework/unsupervised_exploration.py:38]: Just for documentation purposes because I also stumpled across. Let's add comment to clarify why we have k + 1 here: "Query the KDTree for distances. Why k+1 neighbors? Usually, one might exclude the state itself if it’s in the visited set."

**Review by bkuen — DISMISSED** [review]
The implementation of unsupervised pre-training looks absolutely right to me. I have made a few small comments for the purpose of documentation. You can decide for yourself whether you want to include them or merge the PR directly.
And regarding the second replay buffer. This is just a temporary problem which will be automatically resolved if the other PRs of this week are merged as well.

**Review by bkuen — APPROVED** [review]
